File: typescript/searchLinear.ts
/* Linear Search. */

import * as readline from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";

function linearSearch(givenSequence: string, searchTerm: string) {
  /* Linear Search Function */
  const lengthOfSearchTerm = searchTerm.length,
    lengthOfGivenSequence = givenSequence.length;
  let leftCursor = 0,
    rightCursor = lengthOfSearchTerm;

  while (rightCursor <= lengthOfGivenSequence) {
    if (givenSequence.slice(leftCursor, rightCursor) == searchTerm) return true;
    leftCursor += 1;
    rightCursor = leftCursor + lengthOfSearchTerm;
  }
  return false;
}

async function main() {
  /* Main. */
  const sampleInputSequence = "This is an example sequence.",
    sampleInputSearchTerm = "an ex";

  const rl = readline.createInterface({ input, output });

  let inputSequence = (await rl.question("Sequence: ")).trim();
  let inputSearchTerm = (await rl.question("Search term: ")).trim();

  if (inputSequence.length == 0) inputSequence = sampleInputSequence;
  if (inputSearchTerm.length == 0) inputSearchTerm = sampleInputSearchTerm;

  rl.close();

  console.log(linearSearch(inputSequence, inputSearchTerm));
}

main();

File: python/search_linear.py
"""Linear Search."""


def linear_search(given_sequence: str, search_term: str):
    """Linear Search Function."""
    length_of_search_term = len(search_term)
    length_of_given_sequence = len(given_sequence)
    left_cursor = 0
    right_cursor = length_of_search_term

    while right_cursor <= length_of_given_sequence:
        if given_sequence[left_cursor:right_cursor] == search_term:
            return True
        left_cursor += 1
        right_cursor = left_cursor + length_of_search_term
    return False


def main():
    """Main."""
    sample_input_sequence = "This is an example sequence."
    sample_input_search_term = "an ex"

    input_sequence = input("Sequence: ") or sample_input_sequence
    input_search_term = input("Search term: ") or sample_input_search_term

    print(linear_search(input_sequence, input_search_term))


main()
